./oceanus.sh が失敗する

Tippeacanoe がビルドできない？

```
=> ERROR [4/9] RUN git clone https://github.com/mapbox/tippecanoe.git   204.2s 
------                                                                          
 > [4/9] RUN git clone https://github.com/mapbox/tippecanoe.git &&     cd tippecanoe &&     make -j &&     make install:
#7 0.257 Cloning into 'tippecanoe'...
#7 18.80 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geojson.o geojson.cpp
#7 18.80 cc -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -c -o jsonpull/jsonpull.o jsonpull/jsonpull.c
#7 18.80 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o tile.o tile.cpp
#7 18.80 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o pool.o pool.cpp
#7 18.80 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o mbtiles.o mbtiles.cpp
#7 18.80 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geometry.o geometry.cpp
#7 18.80 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o projection.o projection.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o memfile.o memfile.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o mvt.o mvt.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o serial.o serial.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o main.o main.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o text.o text.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o dirtiles.o dirtiles.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o plugin.o plugin.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o read_json.o read_json.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o write_json.o write_json.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geobuf.o geobuf.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o evaluator.o evaluator.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geocsv.o geocsv.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o csv.o csv.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geojson-loop.o geojson-loop.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o enumerate.o enumerate.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o decode.o decode.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o tile-join.o tile-join.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o unit.o unit.cpp
#7 18.81 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o jsontool.o jsontool.cpp
#7 19.63 g++  -L/usr/local/lib -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -o tippecanoe-enumerate enumerate.o  -lsqlite3
#7 21.67 mbtiles.cpp: In function 'void tilestats(const std::map<std::__cxx11::basic_string<char>, layermap_entry>&, size_t, json_writer&)':
#7 21.67 mbtiles.cpp:133:7: warning: variable 'first' set but not used [-Wunused-but-set-variable]
#7 21.67   bool first = true;
#7 21.67        ^~~~~
#7 22.10 mbtiles.cpp: In function 'void mbtiles_write_metadata(sqlite3*, const char*, const char*, int, int, double, double, double, double, double, double, int, const char*, const std::map<std::__cxx11::basic_string<char>, layermap_entry>&, bool, const char*, bool, const std::map<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >&, const string&, const string&)':
#7 22.10 mbtiles.cpp:512:9: warning: variable 'first' set but not used [-Wunused-but-set-variable]
#7 22.10     bool first = true;
#7 22.10          ^~~~~
#7 92.39 g++: internal compiler error: Killed (program cc1plus)
#7 92.39 Please submit a full bug report,
#7 92.39 with preprocessed source if appropriate.
#7 92.39 See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
#7 92.40 Makefile:74: recipe for target 'tile.o' failed
#7 92.40 make: *** [tile.o] Error 4
#7 92.40 make: *** Waiting for unfinished jobs....
#7 163.7 g++: internal compiler error: Killed (program cc1plus)
#7 163.7 Please submit a full bug report,
#7 163.7 with preprocessed source if appropriate.
#7 163.7 See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
#7 163.7 Makefile:74: recipe for target 'unit.o' failed
#7 163.7 make: *** [unit.o] Error 4
------
executor failed running [/bin/sh -c git clone https://github.com/mapbox/tippecanoe.git &&     cd tippecanoe &&     make -j &&     make install]: exit code: 2
cp ./shp2geojson.yaml /tmp
Unable to find image 'geolonia/oceanus:latest' locally
docker: Error response from daemon: pull access denied for geolonia/oceanus, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
docker run -it --rm --name oceanus -v /tmp:/data geolonia/oceanus /app/ne2mbtiles
```

依然ビルドできない状態です。
私の環境は以下の通りです。

- macOS Big Sur 11.0.1
- Docker Desktop 3.0.3

```shell
$ docker -v
Docker version 20.10.0, build 7287ab3
```

```shell
./oceanus.sh 
Building Oceanus Docker Image.Tag name is geolonia/oceanus.
[+] Building 311.3s (8/13)                                                                                                                         
 => [internal] load build definition from Dockerfile                                                                                          0.0s
 => => transferring dockerfile: 37B                                                                                                           0.0s
 => [internal] load .dockerignore                                                                                                             0.0s
 => => transferring context: 2B                                                                                                               0.0s
 => [internal] load metadata for docker.io/library/ubuntu:18.04                                                                               2.5s
 => [internal] load build context                                                                                                             0.0s
 => => transferring context: 3.10kB                                                                                                           0.0s
 => [1/9] FROM docker.io/library/ubuntu:18.04@sha256:ea188fdc5be9b25ca048f1e882b33f1bc763fb976a8a4fea446b38ed0efcbeba                         0.0s
 => CACHED [2/9] RUN apt-get update &&     apt-get -y upgrade &&     apt-get install -y     curl     git     bc     make     rename     pyth  0.0s
 => CACHED [3/9] RUN git clone git://github.com/mapbox/mbutil.git &&     cd mbutil &&     python setup.py install                             0.0s
 => ERROR [4/9] RUN git clone https://github.com/mapbox/tippecanoe.git &&     cd tippecanoe &&     make -j &&     make install              308.7s
------                                                                                                                                             
 > [4/9] RUN git clone https://github.com/mapbox/tippecanoe.git &&     cd tippecanoe &&     make -j &&     make install:                           
#7 0.255 Cloning into 'tippecanoe'...                                                                                                              
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std
=c++11 -c -o geojson.o geojson.cpp                                                                                                                 
#7 16.99 cc -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -c -o
 jsonpull/jsonpull.o jsonpull/jsonpull.c
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o tile.o tile.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o pool.o pool.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o mbtiles.o mbtiles.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geometry.o geometry.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o projection.o projection.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o memfile.o memfile.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o mvt.o mvt.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o serial.o serial.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o main.o main.cpp
#7 16.99 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o text.o text.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o dirtiles.o dirtiles.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o plugin.o plugin.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o read_json.o read_json.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o write_json.o write_json.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geobuf.o geobuf.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o evaluator.o evaluator.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geocsv.o geocsv.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o csv.o csv.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o geojson-loop.o geojson-loop.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o enumerate.o enumerate.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o decode.o decode.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o tile-join.o tile-join.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o unit.o unit.cpp
#7 17.00 g++ -MMD  -I/usr/local/include -I. -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -c -o jsontool.o jsontool.cpp
#7 17.41 g++  -L/usr/local/lib -g -Wall -Wshadow -Wsign-compare -Wextra -Wunreachable-code -Wuninitialized -Wshadow -O3 -DNDEBUG  -std=c++11 -o tippecanoe-enumerate enumerate.o  -lsqlite3
#7 18.38 mbtiles.cpp: In function 'void tilestats(const std::map<std::__cxx11::basic_string<char>, layermap_entry>&, size_t, json_writer&)':
#7 18.38 mbtiles.cpp:133:7: warning: variable 'first' set but not used [-Wunused-but-set-variable]
#7 18.38   bool first = true;
#7 18.38        ^~~~~
#7 18.43 mbtiles.cpp: In function 'void mbtiles_write_metadata(sqlite3*, const char*, const char*, int, int, double, double, double, double, double, double, int, const char*, const std::map<std::__cxx11::basic_string<char>, layermap_entry>&, bool, const char*, bool, const std::map<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >&, const string&, const string&)':
#7 18.43 mbtiles.cpp:512:9: warning: variable 'first' set but not used [-Wunused-but-set-variable]
#7 18.43     bool first = true;
#7 18.43          ^~~~~
#7 63.09 g++: internal compiler error: Killed (program cc1plus)
#7 63.09 Please submit a full bug report,
#7 63.09 with preprocessed source if appropriate.
#7 63.09 See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
#7 63.09 Makefile:74: recipe for target 'mvt.o' failed
#7 63.09 make: *** [mvt.o] Error 4
#7 63.09 make: *** Waiting for unfinished jobs....
#7 210.1 g++: internal compiler error: Killed (program cc1plus)
#7 210.2 Please submit a full bug report,
#7 210.2 with preprocessed source if appropriate.
#7 210.2 See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
#7 210.2 make: *** [tile-join.o] Error 4
#7 210.2 Makefile:74: recipe for target 'tile-join.o' failed
#7 281.0 g++: internal compiler error: Killed (program cc1plus)
#7 281.0 Please submit a full bug report,
#7 281.0 with preprocessed source if appropriate.
#7 281.0 See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
#7 281.0 Makefile:74: recipe for target 'tile.o' failed
#7 281.0 make: *** [tile.o] Error 4
------
executor failed running [/bin/sh -c git clone https://github.com/mapbox/tippecanoe.git &&     cd tippecanoe &&     make -j &&     make install]: exit code: 2
cp ./shp2geojson.yaml /tmp
Unable to find image 'geolonia/oceanus:latest' locally
docker: Error response from daemon: pull access denied for geolonia/oceanus, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
docker run -it --rm --name oceanus -v /tmp:/data geolonia/oceanus /app/ne2mbtiles
```

参考までに僕も同じようなエラーが出ました。

```
mbtiles.cpp: In function 'void tilestats(const std::map<std::__cxx11::basic_string<char>, layermap_entry>&, size_t, json_writer&)':
mbtiles.cpp:133:7: warning: variable 'first' set but not used [-Wunused-but-set-variable]
  bool first = true;
       ^~~~~
mbtiles.cpp: In function 'void mbtiles_write_metadata(sqlite3*, const char*, const char*, int, int, double, double, double, double, double, double, int, const char*, const std::map<std::__cxx11::basic_string<char>, layermap_entry>&, bool, const char*, bool, const std::map<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >&, const string&, const string&)':
mbtiles.cpp:512:9: warning: variable 'first' set but not used [-Wunused-but-set-variable]
    bool first = true;
         ^~~~~
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -20,9 +20,10 @@ RUN git clone git://github.com/mapbox/mbutil.git && \
     cd mbutil && \
     python setup.py install
 
+# The retry of `make -j` is a temporary fix. see https://github.com/geolonia/oceanus/issues/41
 RUN git clone https://github.com/mapbox/tippecanoe.git && \
     cd tippecanoe && \
-    make -j && \
+    (make -j || make -j) && \
     make install    
 
 RUN apt-get install -y python3 \

